firmware/ethereum.c

```diff
@@ -0,0 +1,545 @@
+	}
+
+	layoutDialogSwipe(&bmp_icon_question,
+		"Cancel",
+		"Confirm",
+		NULL,
+		"Send",
+		amount,
+		_to1,
+		_to2,
+		_to3,
+		NULL
+	);
+}
+
+static void layoutEthereumData(const uint8_t *data, uint32_t len, uint32_t total_len)
+{
+	char hexdata[3][17];
+	char summary[20];
+	int i;
+	uint32_t printed = 0;
+	for (i = 0; i < 3; i++) {
+		uint32_t linelen = len - printed;
+		if (linelen > 8)
+			linelen = 8;
+		data2hex(data, linelen, hexdata[i]);
+		data += linelen;
+		printed += linelen;
+	}
+
+	strcpy(summary, "...          bytes");
+	char *p = summary + 11;
+	uint32_t number = total_len;
+	while (number > 0) {
+		*p-- = '0' + number % 10;
+		number = number / 10;
+	}
+	char *summarystart = summary;
+	if (total_len == printed)
+		summarystart = summary + 4;
+
+	layoutDialogSwipe(&bmp_icon_question,
+		"Cancel",
+		"Confirm",
+		NULL,
+		"Transaction data:",
+		hexdata[0],
+		hexdata[1],
+		hexdata[2],
+		summarystart,
+		NULL
+	);
+}
+
+static void layoutEthereumFee(const uint8_t *value, uint32_t value_len,
+							  const uint8_t *gas_price, uint32_t gas_price_len,
+							  const uint8_t *gas_limit, uint32_t gas_limit_len)
+{
+	bignum256 val, gas;
+	uint8_t pad_val[32];
+	char tx_value[25];
+	char gas_value[25];
+
+	memset(pad_val, 0, sizeof(pad_val));
+	memcpy(pad_val + (32 - gas_price_len), gas_price, gas_price_len);
+	bn_read_be(pad_val, &val);
+
+	memset(pad_val, 0, sizeof(pad_val));
+	memcpy(pad_val + (32 - gas_limit_len), gas_limit, gas_limit_len);
+	bn_read_be(pad_val, &gas);
+	bn_multiply(&val, &gas, &secp256k1.prime);
+
+	ethereumFormatAmount(&gas, gas_value);
+
+	memset(pad_val, 0, sizeof(pad_val));
+	memcpy(pad_val + (32 - value_len), value, value_len);
+	bn_read_be(pad_val, &val);
+
+	if (bn_is_zero(&val)) {
+		strcpy(tx_value, "message");
+	} else {
+		ethereumFormatAmount(&val, tx_value);
+	}
+
+	layoutDialogSwipe(&bmp_icon_question,
+		"Cancel",
+		"Confirm",
+		NULL,
+		"Really send",
+		tx_value,
+		"paying up to",
+		gas_value,
+		"for gas?",
+		NULL
+	);
+}
+
+/*
+ * RLP fields:
+ * - nonce (0 .. 32)
+ * - gas_price (0 .. 32)
+ * - gas_limit (0 .. 32)
+ * - to (0, 20)
+ * - value (0 .. 32)
+ * - data (0 ..)
+ */
+
+static bool ethereum_signing_check(EthereumSignTx *msg)
+{
+	if (!msg->has_nonce || !msg->has_gas_price || !msg->has_gas_limit) {
+		return false;
+	}
+	
+	if (msg->to.size != 20 && msg->to.size != 0) {
+		/* Address has wrong length */
+		return false;
+	}
+
+	// sending transaction to address 0 (contract creation) without a data field
+	if (msg->to.size == 0 && (!msg->has_data_length || msg->data_length == 0)) {
+		return false;
+	}
+
+	if (msg->gas_price.size + msg->gas_limit.size  > 30) {
+		// sanity check that fee doesn't overflow
+		return false;
+	}
+
+	return true;
+}
+
+void ethereum_signing_init(EthereumSignTx *msg, const HDNode *node)
+{
+	ethereum_signing = true;
+	sha3_256_Init(&keccak_ctx);
+
+	memset(&resp, 0, sizeof(EthereumTxRequest));
+	/* set fields to 0, to avoid conditions later */
```

Comment:
Please update the comment here which explains that the upper Protobuf layers should do this, but ensure it in the case they didn't.